### PR TITLE
UCT/API/TEST: Stub implementation of uct_iface progress control.

### PR DESCRIPTION
**Pull request by yosefe**

### CODE REVIEW

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3952/ for details (Mellanox internal link).

File: src/uct/api/uct.h [hunk 1]
```diff
@@ -2094,6 +2056,50 @@ UCT_INLINE_API ucs_status_t uct_iface_tag_recv_cancel(uct_iface_h iface,
...
+}
+
+
+/**
+ * @ingroup UCT_RESOURCE
+ * @brief Disable synchronous progress for the interface
+ *
+ * Notify the transport that it should avoid doing anything
+ * during @ref uct_worker_progress(). Thus the latency of
+ * other transports may be reduced.
```
Comment (shamisp) [inline src/uct/api/uct.h]: may be improved

File: src/uct/api/uct.h [hunk 2]
```diff
@@ -2094,6 +2056,50 @@ UCT_INLINE_API ucs_status_t uct_iface_tag_recv_cancel(uct_iface_h iface,
...
+{
+    iface->ops.iface_progress_enable(iface, flags);
+}
+
+
+/**
+ * @ingroup UCT_RESOURCE
+ * @brief Disable synchronous progress for the interface
+ *
+ * Notify the transport that it should avoid doing anything
```
Comment (shamisp) [inline src/uct/api/uct.h]: Avoid is kinda vague ...

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3957/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3958/ for details (Mellanox internal link).

**Comment by MattBBaker** [comment]
```
[       OK ] ud/test_ucp_wireup.send_disconnect_onesided_wait/0 (42 ms)
[ RUN      ] ud/test_ucp_wireup.disconnect_reconnect/1
[pod9:3496 :0] Caught signal 11 (Segmentation fault: address not mapped to object at address 0x33)
[pod9:3496 :0] Process frozen...
make: *** [test] Terminated
Build step 'Execute shell' marked build as failure
```

**Comment by MattBBaker** [comment]
Or that last one was based on a back logged check and this one worked. :+1:

**Comment by yosefe** [comment]
@shamisp @bbenton pls take a look

File: src/uct/api/uct.h [hunk 3]
```diff
@@ -2094,6 +2056,45 @@ UCT_INLINE_API ucs_status_t uct_iface_tag_recv_cancel(uct_iface_h iface,
...
+ * @ingroup UCT_RESOURCE
+ * @brief Enable synchronous progress for the interface
+ *
+ * Notify the transport that it should actively progress communications during
+ * @ref uct_worker_progress().
+ *
+ * When the interface is created, its progress is enabled.
+ *
+ * @param [in]  iface    The interface to enable progress.
+ * @param [in]  flags    What kind progress to enable, see @ref uct_progress_types.
```
Comment (shamisp) [inline src/uct/api/uct.h]: the type of progress to enable as defined by  @ref uct_progress_types.

File: src/uct/api/uct.h [hunk 4]
```diff
@@ -2094,6 +2056,45 @@ UCT_INLINE_API ucs_status_t uct_iface_tag_recv_cancel(uct_iface_h iface,
...
+ * @brief Disable synchronous progress for the interface
+ *
+ * Notify the transport that it should not progress its communications during
+ * @ref uct_worker_progress(). Thus the latency of other transports may be
+ * improved.
+ *
+ * By default, progress is enabled when the interface is created.
+ *
+ * @param [in]  iface    The interface to disable progress.
+ * @param [in]  flags    What kind of progress to disable, see @ref uct_progress_types.
```
Comment (shamisp) [inline src/uct/api/uct.h]: the type of progress to disable as defined by  @ref uct_progress_types.

File: src/uct/cuda/cuda_iface.c [hunk 5]
```diff
@@ -93,6 +93,8 @@ static uct_iface_ops_t uct_cuda_iface_ops = {
     .ep_destroy          = UCS_CLASS_DELETE_FUNC_NAME(uct_cuda_ep_t),
     .iface_flush         = uct_base_iface_flush,
     .iface_fence         = uct_base_iface_fence,
+    .iface_progress_enable = ucs_empty_function,
+    .iface_progress_disable = ucs_empty_function,
```
Comment (shamisp) [inline src/uct/cuda/cuda_iface.c]: indentation in respect to other ```=```

File: src/uct/sm/cma/cma_iface.c [hunk 6]
```diff
@@ -78,6 +78,8 @@ static uct_iface_ops_t uct_cma_iface_ops = {
     .ep_destroy          = UCS_CLASS_DELETE_FUNC_NAME(uct_cma_ep_t),
     .iface_flush         = uct_base_iface_flush,
     .iface_fence         = uct_sm_iface_fence,
+    .iface_progress_enable = ucs_empty_function,
+    .iface_progress_disable = ucs_empty_function,
```
Comment (shamisp) [inline src/uct/sm/cma/cma_iface.c]: indentation

File: src/uct/sm/knem/knem_iface.c [hunk 7]
```diff
@@ -63,6 +63,8 @@ static uct_iface_ops_t uct_knem_iface_ops = {
     .ep_create_connected = UCS_CLASS_NEW_FUNC_NAME(uct_knem_ep_t),
     .ep_destroy          = UCS_CLASS_DELETE_FUNC_NAME(uct_knem_ep_t),
     .iface_fence         = uct_sm_iface_fence,
+    .iface_progress_enable = ucs_empty_function,
+    .iface_progress_disable = ucs_empty_function,
```
Comment (shamisp) [inline src/uct/sm/knem/knem_iface.c]: indentation

File: src/uct/sm/mm/mm_iface.c [hunk 8]
```diff
@@ -273,6 +273,8 @@ static uct_iface_ops_t uct_mm_iface_ops = {
     .ep_destroy          = UCS_CLASS_DELETE_FUNC_NAME(uct_mm_ep_t),
     .iface_flush         = uct_mm_iface_flush,
     .iface_fence         = uct_sm_iface_fence,
+    .iface_progress_enable = ucs_empty_function,
+    .iface_progress_disable = ucs_empty_function,
```
Comment (shamisp) [inline src/uct/sm/mm/mm_iface.c]: indentation

File: src/uct/ugni/rdma/ugni_rdma_iface.c [hunk 9]
```diff
@@ -203,10 +203,13 @@ static uct_iface_ops_t uct_ugni_gemini_rdma_iface_ops = {
     .ep_pending_add      = uct_ugni_ep_pending_add,
     .ep_pending_purge    = uct_ugni_ep_pending_purge,
     .ep_flush            = uct_ugni_ep_flush,
+    .ep_fence            = uct_base_ep_fence,
     .ep_create_connected = UCS_CLASS_NEW_FUNC_NAME(uct_ugni_ep_t),
     .ep_destroy          = UCS_CLASS_DELETE_FUNC_NAME(uct_ugni_ep_t),
     .iface_flush         = uct_ugni_iface_flush,
     .iface_fence         = uct_base_iface_fence,
+    .iface_progress_enable = ucs_empty_function,
+    .iface_progress_disable = ucs_empty_function,
```
Comment (shamisp) [inline src/uct/ugni/rdma/ugni_rdma_iface.c]: indentation

File: src/uct/ugni/smsg/ugni_smsg_iface.c [hunk 10]
```diff
@@ -237,6 +237,8 @@ static uct_iface_ops_t uct_ugni_smsg_iface_ops = {
     .ep_connect_to_ep      = uct_ugni_smsg_ep_connect_to_ep,
     .iface_flush           = uct_ugni_iface_flush,
     .iface_fence           = uct_base_iface_fence,
+    .iface_progress_enable = ucs_empty_function,
+    .iface_progress_disable = ucs_empty_function,
```
Comment (shamisp) [inline src/uct/ugni/smsg/ugni_smsg_iface.c]: indentation

File: src/uct/ugni/udt/ugni_udt_iface.c [hunk 11]
```diff
@@ -359,6 +359,8 @@ static uct_iface_ops_t uct_ugni_udt_iface_ops = {
     .ep_destroy            = UCS_CLASS_DELETE_FUNC_NAME(uct_ugni_udt_ep_t),
     .iface_flush           = uct_ugni_iface_flush,
     .iface_fence           = uct_base_iface_fence,
+    .iface_progress_enable = ucs_empty_function,
+    .iface_progress_disable = ucs_empty_function,
```
Comment (shamisp) [inline src/uct/ugni/udt/ugni_udt_iface.c]: indentation

**Comment by mellanox-github** [comment]
Test FAILed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3961/ for details (Mellanox internal link).

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3962/ for details (Mellanox internal link).

**Comment by MattBBaker** [comment]
Crashes on UGNI. 
```
[nid00745:11514:0]   uct_iface.c:314  Assertion `ops->iface_progress_enable != NULL' failed
```
Will poke what is causing.

**Comment by yosefe** [comment]
@MattBBaker latest commit should fix this

**Comment by MattBBaker** [comment]
@yosefe Yes it does.

**Comment by yosefe** [comment]
@bbenton pls take a look

**Comment by mellanox-github** [comment]
Test PASSed.
See http://hpc-master.lab.mtl.com:8080/job/hpc-ucx-pr/3967/ for details (Mellanox internal link).

**Comment by bbenton** [comment]
:+1: for Doc/API

**Pull request merged**